weave/trace/serialization/custom_objs.py

```diff
@@ -56,80 +82,95 @@ def encode_custom_obj(obj: Any) -> dict | None:
+            load_instance_op_ref = wc._save_op(
+                serializer.load, "load_" + serializer.id()
+            )  # type: ignore
+            load_op_uri = load_instance_op_ref.uri()
+
+    encoded: EncodedCustomObjDict = {
         "_type": "CustomWeaveType",
         "weave_type": {"type": serializer.id()},
         "load_op": load_op_uri,
     }
 
-    # If the save method just takes one argument, it is an inline serializer
-    if is_inline_save(serializer.save):
-        encoded["val"] = serializer.save(obj)
-    elif is_file_save(serializer.save):
-        art = MemTraceFilesArtifact()
-        serializer.save(obj, art, "obj")
+    art = MemTraceFilesArtifact()
+    val = serializer.save(obj, art, "obj")
+    if art.path_contents:
         encoded_path_contents = {
             k: (v.encode("utf-8") if isinstance(v, str) else v)  # type: ignore
             for k, v in art.path_contents.items()
         }
         encoded["files"] = encoded_path_contents
-    else:
-        raise ValueError(
-            f"Serializer save function could not be identified as inline or file-based: {type(serializer.save)}"
-        )
+    if val is not None:
+        encoded["val"] = val
+
     return encoded
 
 
-def decode_custom_inline_obj(obj: dict) -> Any:
-    type_ = obj["weave_type"]["type"]
-    if type_ in KNOWN_TYPES:
-        serializer = get_serializer_by_id(type_)
-        if serializer is not None:
-            if is_op(serializer.load):
-                # We would expect this to be already set to False, but
-                # just in case.
-                serializer.load._tracing_enabled = False  # type: ignore
-            return serializer.load(obj["val"])
-
-    load_op_uri = obj.get("load_op")
-    if load_op_uri is None:
-        raise ValueError(f"No serializer found for `{type_}`")
-
-    op_ref = OpRef.parse_uri(load_op_uri)
-    wc = require_weave_client()
-    load_instance_op = wc.get(op_ref)
-    if load_instance_op is None:
-        raise ValueError(
-            f"Failed to load op needed to decode object of type `{type_}`. See logs above for more information."
+def _load_custom_obj_files(
+    project_id: str, server: TraceServerInterface, file_digests: dict
+) -> dict[str, bytes]:
+    loaded_files: dict[str, bytes] = {}
+    for name, digest in file_digests.items():
+        file_response = server.file_content_read(
+            FileContentReadReq(project_id=project_id, digest=digest)
         )
+        loaded_files[name] = file_response.content
+    return loaded_files
 
-    load_instance_op._tracing_enabled = False  # type: ignore
-    return load_instance_op(obj.get("val"))
+
+def maybe_attach_art(res: Any, art: MemTraceFilesArtifact) -> None:
+    """Attach the artifact to the result if possible.
+    This is a best-effort approach to attach the artifact to the result.
+
+    I am pretty sure this is just used in a legacy code capture case
+    """
+    try:
+        res.art = art
+    except Exception:
```

Comment:
Could we log the error here when one occurs? This line actually caused a massive headache while working on content and was quite a pain to track down.